codalab/worker/download_util.py

```diff
@@ -125,9 +175,118 @@ def _compute_target_info(path, depth):
         result['type'] = 'directory'
         if depth > 0:
             result['contents'] = [
-                _compute_target_info(os.path.join(path, file_name), depth - 1)
+                _compute_target_info_local(os.path.join(path, file_name), depth - 1)
                 for file_name in os.listdir(path)
             ]
     if result is None:
         raise PathException()
     return result
+
+
+def _compute_target_info_beam(path: str, depth: Union[int, float]) -> TargetInfo:
+    """Computes target info for a file that is externalized on a location
+    such as Azure, by using the Apache Beam FileSystem APIs."""
+
+    linked_bundle_path = parse_linked_bundle_url(path)
+    if not FileSystems.exists(linked_bundle_path.bundle_path):
+        raise PathException(linked_bundle_path.bundle_path)
+    if not linked_bundle_path.is_archive:
+        # Single file
+        raise PathException(
+            "Single files on Blob Storage are not supported; only a path within a .tar.gz fils is supported."
+        )
+
+    with OpenIndexedTarGzFile(linked_bundle_path.bundle_path) as tf:
+        islink = lambda finfo: stat.S_ISLNK(finfo.mode)
+        readlink = lambda finfo: finfo.linkname
+
+        isfile = lambda finfo: finfo.type in tarfile.REGULAR_TYPES
+        isdir = lambda finfo: finfo.type == tarfile.DIRTYPE
+        listdir = lambda path: tf.getFileInfo(path, listDir=True)
+
+        def _get_info(path: str, depth: Union[int, float]) -> TargetInfo:
+            """This function is called to get the target info of the specified path.
+            If the specified path is a directory and additional depth is requested, this
+            function is recursively called to retrieve the target info of files within
+            the directory, much like _compute_target_info_local.
+            """
+            if not path.startswith("/"):
```

Comment:
can these paths be relative?

if path is `foo.txt`, it would look for the file in the root directory?

Reply:
Due to the way ratarmount parses paths, `foo.txt` would return no results, but `/foo.txt` refers to a relative path that would look for the file in the root directory (within the tar file, of course).